### PR TITLE
Make RxJava's RequestContext propagation put context when subscriber is created instead

### PR DESCRIPTION
**Pull request by kojilin**
**Motivation:**
* #3051
* ~This PR is WIP, need to test most cases.~

**Modifications:**

* Using RxJavaPlugin's setOnXxxSubscribe instead setOnXxxAssembly to propagate context when subscribing 
* Add few test cases for testing `Single#cache` and `Flowable#publish`. The workaround is acceptable? Because it's not common to use these cases.
* Using ServiceRequestContext to test `push` strictly.

**TODO in this PR:**
- [x] Add more test cases like `zip`, `retryWhen`, etc. 

**Result:**
* Make the current hook works like other instrument libraries. This will break the existing behavior, but works naturally.

### CODE REVIEW

File: rxjava2/src/main/java/com/linecorp/armeria/common/rxjava2/RequestContextAssembly.java [hunk 1]
```diff
@@ -16,65 +16,58 @@
...
-    @GuardedBy("RequestContextAssembly.class")
-    private static Function<? super Observable, ? extends Observable> oldOnObservableAssembly;
     @SuppressWarnings("rawtypes")
     @Nullable
     @GuardedBy("RequestContextAssembly.class")
-    private static Function<? super ConnectableObservable, ? extends ConnectableObservable>
-            oldOnConnectableObservableAssembly;
+    private static BiFunction<? super Observable, ? super Observer, ? extends Observer>
+            oldOnObservableSubscribe;
     @SuppressWarnings("rawtypes")
```
Comment (minwoox) [inline rxjava2/src/main/java/com/linecorp/armeria/common/rxjava2/RequestContextAssembly.java:51]: nit: we can now remove this line

**Review by minwoox — COMMENTED** [review]
Looks good! Let's keep working on it. 😄

**Comment by trustin** [comment]
@kojilin Ready for reviews?

**Comment by kojilin** [comment]
Will add more test & finish this week.
I think my test code is quite redundant, if there is any suggestion..please^^;;

File: rxjava2/src/main/java/com/linecorp/armeria/common/rxjava2/RequestContextSubscriber.java [hunk 2]
```diff
@@ -69,4 +69,18 @@ public int requestFusion(int mode) {
     public T poll() throws Exception {
         return qs.poll();
     }
+
+    @Override
+    public void request(long n) {
```
Comment (kojilin) [inline rxjava2/src/main/java/com/linecorp/armeria/common/rxjava2/RequestContextSubscriber.java:74]: I think we need this to support backpressure awarded Flowable.

**Comment by kojilin** [comment]
Updated. I think it's ready for review.

**Comment by kojilin** [comment]
retest please

File: rxjava2/src/main/java/com/linecorp/armeria/common/rxjava2/RequestContextAssembly.java [hunk 3]
```diff
@@ -16,65 +16,59 @@
...
-import io.reactivex.observables.ConnectableObservable;
-import io.reactivex.parallel.ParallelFlowable;
+import io.reactivex.SingleObserver;
+import io.reactivex.functions.BiFunction;
+import io.reactivex.internal.fuseable.ConditionalSubscriber;
 import io.reactivex.plugins.RxJavaPlugins;
 
 /**
  * Utility class to keep {@link RequestContext} during RxJava operations.
+ * This plugin doesn't support Flowable#parallel.
```
Comment (trustin) [inline rxjava2/src/main/java/com/linecorp/armeria/common/rxjava2/RequestContextAssembly.java:44]: How about using `{@link ..}` for `Flowable#parallel`?

File: rxjava2/src/main/java/com/linecorp/armeria/common/rxjava2/RequestContextAssembly.java [hunk 4]
```diff
@@ -16,65 +16,59 @@
...
+ * https://github.com/ReactiveX/RxJava/issues/7190
```
Comment (trustin) [inline rxjava2/src/main/java/com/linecorp/armeria/common/rxjava2/RequestContextAssembly.java:45]: How about `See <a href="...">ReactiveX/RxJava#7190</a> for more information.`?

**Review by trustin — APPROVED** [review]
This is really nice. Do you think we can do the same for Reactor, @minwoox ?

**Comment by minwoox** [comment]
> This is really nice. Do you think we can do the same for Reactor, @minwoox ?

Yes. Let me send the follow-up PR. 😄

**Review by minwoox — APPROVED** [review]
Thanks a lot, @kojilin!

File: rxjava2/src/main/java/com/linecorp/armeria/common/rxjava2/RequestContextAssembly.java [hunk 5]
```diff
@@ -91,161 +85,120 @@ public static synchronized void enable() {
...
+                oldOnObservableSubscribe,
+                new BiConditionalOnCurrentRequestContextFunction<Observable, Observer>() {
                     @Override
-                    Observable applyActual(Observable o, RequestContext ctx) {
-                        if (!(o instanceof Callable)) {
-                            return new RequestContextObservable(o, ctx);
+                    Observer applyActual(Observable observable, Observer observer, RequestContext ctx) {
+                        try (SafeCloseable ignored = ctx.push()) {
+                            return new RequestContextObserver<>(observer, ctx);
                         }
```
Comment (ikhoon) [inline rxjava2/src/main/java/com/linecorp/armeria/common/rxjava2/RequestContextAssembly.java:96]: Don't we need to push a `RequestContext` while creating a `RequestContextObserver`?

File: rxjava3/src/main/java/com/linecorp/armeria/common/rxjava3/RequestContextSubscriber.java [hunk 6]
```diff
@@ -69,4 +69,18 @@ public int requestFusion(int mode) {
     public T poll() throws Throwable {
         return qs.poll();
     }
+
+    @Override
+    public void request(long n) {
```
[the remark this answers is not in the document]
Comment (kojilin) [reply rxjava3/src/main/java/com/linecorp/armeria/common/rxjava3/RequestContextSubscriber.java:74]: I think maybe same as ur another comment, beforeDownstream & afterDownstream donesn't touch the downstream. And it's used like
```
            if (beforeDownstream()) {

                downstream.onSubscribe(this);

                afterDownstream();
            }
```
And we already hook onSubscribe with
```
        @Override
        public void onSubscribe(Subscription s) {
            try (SafeCloseable ignored = ctx.push()) {
                subscriber.onSubscribe(s);
            }
        }
```

so maybe we don't need it,  WDYT?

**Comment by kojilin** [comment]
Updated, remove unnecessary context push at RequestContextAssembly

File: rxjava3/src/main/java/com/linecorp/armeria/common/rxjava3/RequestContextSubscriber.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ikhoon) [reply rxjava3/src/main/java/com/linecorp/armeria/common/rxjava3/RequestContextSubscriber.java:74]: The `onSubscribe()` would be called sequentially on RxJava observer chains.
However the context is pushed on `RequestContextMaybeObserver.onSubscribe()`.
https://github.com/line/armeria/blob/fbe90f0f2a952578012a552c26b3f4f8bdb33c8c/rxjava3/src/main/java/com/linecorp/armeria/common/rxjava3/RequestContextMaybeObserver.java#L36-L45
But, `RequestContextObserver` does nothing on `onSubscribe()`. 🤔
> And we already hook onSubscribe with

Does this mean we don't need to push a context for a `RequestContextObserver`?

[the remark this answers is not in the document]
Comment (kojilin) [reply rxjava3/src/main/java/com/linecorp/armeria/common/rxjava3/RequestContextSubscriber.java:92]: You are right, I'm thinking why our test works and think we already do at `onSubscribe`. But it's final ^^;;.

So maybe subscribe & onSubscribe hapeen at same thread in my test, so it looks like onSubscribe has ctx even I didn't add the hook code to before/after DownStream. But looks like it's better to add to follow the pattern like other Single/Maybe/Completable's onSubscribe.

[the remark this answers is not in the document]
Comment (kojilin) [reply rxjava3/src/main/java/com/linecorp/armeria/common/rxjava3/RequestContextSubscriber.java:92]: Also curious if there is a case that is it always having the same thread in `onSubscribe` and when calling`subscribe`.

**Comment by kojilin** [comment]
updated. Add ctx push/close code at beforeDownstream and afterDownstream.

**Comment by kojilin** [comment]
Please wait for a while, I wanna check few edge cases^^.

**Comment by kojilin** [comment]
updated, maybe it's become simpler.

**Comment by kojilin** [comment]
updated, maybe it's become simpler.

File: rxjava3/src/main/java/com/linecorp/armeria/common/rxjava3/RequestContextSubscriber.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ikhoon) [reply rxjava3/src/main/java/com/linecorp/armeria/common/rxjava3/RequestContextSubscriber.java:92]: > it always having the same thread in onSubscribe and when callingsubscribe.

As per [Reactive Stream specification 1.3](https://github.com/reactive-streams/reactive-streams-jvm#1.3), if happen-before relation is guaranteed between signals, `onSubscribe()` could be called in multi-thread environment.

**Review by ikhoon — APPROVED** [review]
Thanks! @kojilin

**Pull request closed without merging**